README.md
# Dailyline

Calculate your daily programming time and number of lines

## Demo

1.command(ctrl) + shift + p

> dailyline

2.click statusBar

![2020-04-18-23.19.21](https://s3.qiufengh.com/blog/2020-04-18-23.19.21.gif)
## config

**dailyline.questionConfig**

type: Boolean

default: fasle

desc: Ask every day(only Chinese)

```json
...
"dailyline.questionConfig": false
...
```


## Thanks

icon from 

https://www.iconfont.cn/user/detail?spm=a313x.7781069.0.d214f71f6&uid=757066

https://www.iconfont.cn/user/detail?spm=a313x.7781069.0.d214f71f6&uid=53892

CHANGELOG.md
# Change Log

2020.04.18
1.Daily configuration
2.cache clean

2020.04.15 mvp version
1.Count the number of lines of code written
2.Statistics code writing time

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import { StatusBarItem, commands, window, workspace, ExtensionContext, ViewColumn, Uri, WebviewPanel } from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import { homedir } from 'os';
import DailyLine from './dailyline';
const daily = new DailyLine();
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: ExtensionContext) {
	let panel: WebviewPanel;
	daily.init(context);
	var debounce = function (func: Function, wait: number, immediate: boolean) {
		// 设置定时器
		let timeout: any;
		return (...args: any[]) => {
			const later = () => {
				timeout = null;
				if (!immediate) func.apply(this, args);
			};
			const callNow = immediate && !timeout;
			clearTimeout(timeout);
			timeout = setTimeout(later, wait);
			if (callNow)
				func.apply(this, args);
		}
	};

	const htmlPath = path.resolve(context.extensionPath, 'webview/index.html');
	let lastUsedImageUri = Uri.file(path.resolve(homedir(), 'Desktop/code.png'));

	workspace.onDidOpenTextDocument(() => {
		daily.open();
	});
	workspace.onDidSaveTextDocument(() => {
		daily.save();
	});

	workspace.onDidChangeTextDocument(debounce(() => {
		console.log('active');
		daily.setNewTimer();
	}, 500, true));

	
	window.registerWebviewPanelSerializer('dailyline', {
		async deserializeWebviewPanel(_panel: WebviewPanel, state: any) {
			panel = _panel;
			daily.setPanel(panel);
			panel.webview.html = getHtmlContent(htmlPath);
			daily.setcode();
			setupMessageListeners();
		}
	});

	commands.registerCommand('dailyline.start', () => {
		// 创建并显示新的webview
		panel = window.createWebviewPanel('dailyline', 'dailyline', 2, {
			enableScripts: true,
			localResourceRoots: [Uri.file(path.join(context.extensionPath, 'webview'))]
		});
		daily.setPanel(panel);
		panel.webview.html = getHtmlContent(htmlPath);
		daily.setcode();
		setupMessageListeners();
	});

	const writeSerializedBlobToFile = (serializeBlob: any, fileName: string) => {
		const bytes = new Uint8Array(serializeBlob.split(','));
		fs.writeFileSync(fileName, Buffer.from(bytes));
	};

	function setupMessageListeners() {
		panel.webview.onDidReceiveMessage(({ type, data }) => {
			switch (type) {
				case 'shoot':
					window
						.showSaveDialog({
							defaultUri: lastUsedImageUri,
							filters: {
								Images: ['png']
							}
						})
						.then(uri => {
							if (uri) {
								writeSerializedBlobToFile(data.serializedBlob, uri.fsPath);
								lastUsedImageUri = uri;
							}
						});
					break;
			}
		});
	}
	context.subscriptions.push(daily);
}

// this method is called when your extension is deactivated
export function deactivate() {
	daily.syncLocal();
}

function getHtmlContent(htmlPath: string) {
	const htmlContent = fs.readFileSync(htmlPath, 'utf-8');
	return htmlContent.replace(/script src="([^"]*)"/g, (match, src) => {
		const realSource = 'vscode-resource:' + path.resolve(htmlPath, '..', src);
		return `script src="${realSource}"`;
	});
}



webview/index.js
(function () {
  const vscode = acquireVsCodeApi();

  let target = "container";
  let transparentBackground = false;
  let backgroundColor = "#f2f2f2";

  vscode.postMessage({
    type: "getAndUpdateCacheAndSettings",
  });

  const snippetNode = document.getElementById("snippet");
  const innerWrapper = document.getElementById("snippet-inner");
  const snippetContainerNode = document.getElementById("snippet-container");
  const obturateur = document.getElementById("save");

  snippetContainerNode.style.opacity = "1";
  const oldState = vscode.getState();
  if (oldState && oldState.innerHTML) {
    snippetNode.innerHTML = oldState.innerHTML;
  }
  var bar_styles = ["░▒▓█"];

  function repeat(s, i) {
    var r = "";
    for (var j = 0; j < i; j++) r += s;
    return r;
  }
  function make_bar(p, bar_style, min_size, max_size) {
    var d,
      full,
      m,
      middle,
      r,
      rest,
      x,
      min_delta = Number.POSITIVE_INFINITY,
      full_symbol = bar_style[bar_style.length - 1],
      n = bar_style.length - 1;
    if (p == 100) return { str: repeat(full_symbol, 10), delta: 0 };
    p = p / 100;
    for (var i = max_size; i >= min_size; i--) {
      x = p * i;
      full = Math.floor(x);
      rest = x - full;
      middle = Math.floor(rest * n);
      if (p != 0 && full == 0 && middle == 0) middle = 1;
      d = Math.abs(p - (full + middle / n) / i) * 100;
      if (d < min_delta) {
        min_delta = d;
        m = bar_style[middle];
        if (full == i) m = "";
        r = repeat(full_symbol, full) + m + repeat(bar_style[0], i - full - 1);
      }
    }
    return { str: r, delta: min_delta };
  }

  const serializeBlob = (blob, cb) => {
    const fileReader = new FileReader();

    fileReader.onload = () => {
      const bytes = new Uint8Array(fileReader.result);
      cb(Array.from(bytes).join(","));
    };
    function getBrightness(color) {
      const rgb = this.toRgb();
      return (rgb.r * 299 + rgb.g * 587 + rgb.b * 114) / 1000;
    }

    fileReader.readAsArrayBuffer(blob);
  };

  function shoot(serializedBlob) {
    vscode.postMessage({
      type: "shoot",
      data: {
        serializedBlob,
      },
    });
  }

  function getBrightness(hexColor) {
    const rgb = parseInt(hexColor.slice(1), 16);
    const r = (rgb >> 16) & 0xff;
    const g = (rgb >> 8) & 0xff;
    const b = (rgb >> 0) & 0xff;
    return (r * 299 + g * 587 + b * 114) / 1000;
  }

  obturateur.addEventListener("click", () => {
    if (target === "container") {
      shootAll();
    } else {
      shootSnippet();
    }
  });

  function shootAll() {
    const width = snippetContainerNode.offsetWidth * 2;
    const height = snippetContainerNode.offsetHeight * 2;
    const config = {
      width,
      height,
      style: {
        transform: "scale(2)",
        "transform-origin": "center",
        background: getRgba(backgroundColor, transparentBackground),
      },
    };

    // Hide resizer before capture
    snippetNode.style.resize = "none";
    snippetContainerNode.style.resize = "none";

    domtoimage.toBlob(snippetContainerNode, config).then((blob) => {
      snippetNode.style.resize = "";
      snippetContainerNode.style.resize = "";
      serializeBlob(blob, (serializedBlob) => {
        shoot(serializedBlob);
      });
    });
  }

  function shootSnippet() {
    const width = snippetNode.offsetWidth * 2;
    const height = snippetNode.offsetHeight * 2;
    const config = {
      width,
      height,
      style: {
        transform: "scale(2)",
        "transform-origin": "center",
        padding: 0,
        background: "none",
      },
    };

    // Hide resizer before capture
    snippetNode.style.resize = "none";
    snippetContainerNode.style.resize = "none";

    domtoimage.toBlob(snippetContainerNode, config).then((blob) => {
      snippetNode.style.resize = "";
      snippetContainerNode.style.resize = "";
      serializeBlob(blob, (serializedBlob) => {
        shoot(serializedBlob);
      });
    });
  }

  function getCodeHtml(line, time, day) {
    let timeHTML = '';
    if (time < 1000 * 60 * 60) {
      timeHTML = Math.fround(time / 1000 / 60).toFixed(2) + 'min';
    } else {
      const hours = time / 1000 / 60 / 60;
      const timeradio = ~~((hours * 100) / 24);
      timeobj = make_bar(timeradio, bar_styles[0], 5, 20);
      timeHTML = `${timeobj.str} ${timeradio}%`;
    }
    const logo = `<div id="logo"><svg width="30px" height="30px" viewBox="0 0 256 256" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" preserveAspectRatio="xMidYMid">
    <g>
        <path d="M191.97885,0 L191.97885,219.867316 L0,191.353848 L191.97885,255.998083 L255.97885,229.374998 L255.97885,30.592308 L256,30.5826932 L255.97885,30.5403853 L255.97885,26.6250001 L191.97885,0 Z M124.796158,37.4576925 L65.9942314,95.5826932 L30.5923079,68.9211553 L16.0019234,73.7942316 L52.0192315,109.398076 L16.0019234,145.000006 L30.5923079,149.875005 L65.9942314,123.211541 L65.9961553,123.211541 L124.794228,181.332699 L160.011538,166.369233 L160.011538,52.4211546 L124.796158,37.4576925 Z M124.794228,78.9307702 L124.794228,139.857695 L84.340386,109.394236 L124.794228,78.9307702 Z" fill="#016EC5"></path>
    </g>
  </svg></div>`;
    return `<div id="day">${day}</div><div>💻 Today coding ${line} lines</div><div>⏱️ ${timeHTML}</div><div class="line"></div>`;
  }

  window.addEventListener("message", (e) => {
    if (e) {
      if (e.data.type === "update") {
        const { fontFamily, bgColor, line, time, day, quesNum, question } = e.data;
        console.log(e.data);
        let defaulHtml = getCodeHtml(line, time, day);
        if (quesNum) {
          const qr = document.getElementById("qr");
          qr.style.display = 'block';
          const qrcode = new QRCode(qr, {
            text: `https://github.com/hua1995116/daily-questions/issues/${quesNum}`,
            width: 60,
            height: 60,
            colorDark : "#000000",
            colorLight : "#ffffff",
            correctLevel : QRCode.CorrectLevel.H
          });
          defaulHtml += `<div>❓ ${question}</div>`;
        }
        innerWrapper.innerHTML = defaulHtml;
      }
    }
  });
})();

function getRgba(hex, transparentBackground) {
  const bigint = parseInt(hex.slice(1), 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;
  const a = transparentBackground ? 0 : 1;
  return `rgba(${r}, ${g}, ${b}, ${a})`;
}










webpack.config.js

"use strict";

const path = require("path");

/**@type {import('webpack').Configuration}*/
const config = {
  target: "node", // vscode插件运行在Node.js环境中 📖 -> https://webpack.js.org/configuration/node/

  entry: "./src/extension.ts", // 插件的入口文件 📖 -> https://webpack.js.org/configuration/entry-context/
  output: {
    // 打包好的文件储存在'dist'文件夹中 (请参考package.json), 📖 -> https://webpack.js.org/configuration/output/
    path: path.resolve(__dirname, "out"),
    filename: "extension.js",
    libraryTarget: "commonjs2",
    // devtoolModuleFilenameTemplate: "../[resource-path]"
  },
  devtool: "source-map",
  externals: {
    vscode: "commonjs vscode" // vscode-module是热更新的临时目录，所以要排除掉。 在这里添加其他不应该被webpack打包的文件, 📖 -> https://webpack.js.org/configuration/externals/
  },
  resolve: {
    // 支持读取TypeScript和JavaScript文件, 📖 -> https://github.com/TypeStrong/ts-loader
    extensions: [".ts", ".js"]
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: "ts-loader"
          }
        ]
      }
    ]
  }
};
module.exports = config;

src/dailyline.ts
import { StatusBarAlignment, window, StatusBarItem, WebviewPanel, ExtensionContext, workspace } from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import axios from 'axios';

interface Cache {
	[key: string]: {
		line: number;
		timesteps: TimeCode[],
		question: string,
		quesNum: number
	}
}

interface DayMap {
	[key: string]: {
		[key: string]: number
	}
}

interface TimeCode {
	startTime: number;
	endTime: number | null;
}
class DailyLine {
	cache: Cache;
	dayMap: DayMap;
	context: ExtensionContext | null;
	cachePath: string;
	statusBar: StatusBarItem;
	htmlPath: string;
	timer: any;
	panel: WebviewPanel | null;
	static INTERVAL = 1000 * 60 * 3;
	constructor() {
		this.context = null;
		this.panel = null;
		this.cache = {};
		this.dayMap = {};
		this.timer = null;
		this.cachePath = '';
		this.htmlPath = '';
		this.statusBar = window.createStatusBarItem(StatusBarAlignment.Left);
    }
    init(context: ExtensionContext) {
        this.context = context;
		this.cachePath = context.globalStoragePath;
		this.htmlPath = path.resolve(context.extensionPath, 'webview/index.html');
		this.initCache();
		const today = this.getToday();
		this.checkInit(today);
		this.statusBar.text = this.setStatusText(this.cache[today].line || 0);
		this.statusBar.show();
		this.statusBar.command = "dailyline.start";
	}
	setStatusText(line: number) {
		return `$(code) ${line} lines`;
	}
	setNewTimer() {
		if (!this.timer) {
			const today = this.getToday()
			this.initCache();
			const { timesteps } = this.cache[today];
			timesteps.push({
				startTime: +new Date(),
				endTime: null
			})
			this.syncLocal();
		}
		if (this.timer) {
			clearTimeout(this.timer);
		}
		this.timer = setTimeout(() => {
			const today = this.getToday()
			const { timesteps } = this.cache[today];
			const lastcode = timesteps[timesteps.length - 1];

			if (!lastcode.endTime) {
				lastcode.endTime = Date.now();
			}

			if ((lastcode.endTime + DailyLine.INTERVAL) < Date.now()) {
				lastcode.endTime = Date.now();
			}

			this.syncLocal();
			if (this.timer) {
				clearTimeout(this.timer);
			}
			this.timer = null;
		}, DailyLine.INTERVAL);
	}
	computedTime(): number {
		const today = this.getToday()
		const { timesteps } = this.cache[today];
		return timesteps.reduce((totalTime, timestep ) => {
			if (timestep.endTime) {
				return (timestep.endTime - timestep.startTime) + totalTime;
			}
			const now = Date.now();
			if ((now - timestep.startTime) < DailyLine.INTERVAL) {
				return now - timestep.startTime + totalTime;
			}
			return DailyLine.INTERVAL + totalTime;
		}, 0);
	}
	setPanel(panel: WebviewPanel) {
		this.panel = panel;
	}
	initCache() {
		if (!fs.existsSync(this.cachePath)) {
			fs.writeFileSync(this.cachePath, JSON.stringify(this.cache));
		} else {
			const content = fs.readFileSync(this.cachePath).toString();
			try {
				const cache = JSON.parse(content);
				const cacheList = Object.keys(cache);
				if (cacheList.length > 7) {
					const deleteList = cacheList.slice(0, cacheList.length - 7);
					deleteList.forEach(item => {
						delete cache[item];
					})
				}
				this.cache = cache;
			} catch (e) {

			}
		}
	}
	getCacheLine() {
		const today = this.getToday();
		return this.cache[today].line;
	}
	getToday() {
		const date = new Date();
		const year = date.getFullYear();
		const month = date.getMonth();
		const day = date.getDate();
		return `${year}-${(month + 1) >= 10 ? (month + 1) : '0' + (month + 1)}-${day >= 10 ? day : '0' + day}`;
	}
	getLine() {
		let editor = window.activeTextEditor;
		if (!editor) {
			return 0;
		}
		const content = editor.document.getText();
		const lastLine = content.split(/\r?\n/);
		return lastLine.length;
	}
	open() {
		let editor = window.activeTextEditor;
		if (!editor) {
			return;
		}
		const fileName = editor.document.fileName;
		if (!fileName) {
			return;
		}
		const today = this.getToday();
		this.checkInit(today, fileName);
	}
	checkInit(today: string, fileName?: string) {
		if (!this.cache[today]) {
			this.cache[today] = {
				line: 0,
				timesteps: [],
				quesNum: 0,
				question: ''
			};
		}
		if (!this.dayMap[today]) {
			this.dayMap[today] = {};
		}
		if (fileName) {
			if (!this.dayMap[today][fileName]) {
				this.dayMap[today][fileName] = this.getLine();
			}
		}
	}
	save() {
		this.initCache();
		const today = this.getToday();
		let editor = window.activeTextEditor;
		if (!editor) {
			return;
		}
		const fileName = editor.document.fileName;
		this.checkInit(today, fileName);
		if (!fileName) {
			return;
		}
		const saveLine = this.getLine();
		const addcode = saveLine - this.dayMap[today][fileName];
		if (addcode > 0) {
			this.dayMap[today][fileName] = saveLine;
			this.cache[today].line += addcode;
			this.syncLocal()
			this.statusBar.text = this.setStatusText(this.cache[today].line);
			if (this.panel) {
				this.setcode();
			}
		}
	}
	syncLocal() {
		fs.writeFileSync(this.cachePath, JSON.stringify(this.cache));
	}
	setcode() {
		const panel = this.panel;
		const line = this.getCacheLine();
		const day = this.getToday();
		const time = this.computedTime();
		const fontFamily = workspace.getConfiguration('editor').fontFamily;
		const _this = this;
		const msg = {
			type: 'update',
			line,
			fontFamily,
			bgColor: '#2e3440',
			time,
			day,
			quesNum: 0,
			question: '',
		};

		const questionConfig = workspace.getConfiguration('dailyline').get('questionConfig');
		if (questionConfig) {
			const todayCache = this.cache[day];
			if (todayCache.quesNum) {
				msg.quesNum = todayCache.quesNum;
				msg.question = todayCache.question;
				console.log(msg, 'cache');
				if (panel) {
					panel.webview.postMessage(msg);
				}
			} else {
				axios.get('https://api.github.com/repos/hua1995116/daily-questions/issues?page=1').then((res: any) => {
					const result = res.data;
					const lastquestion = result[0];
					todayCache.quesNum = lastquestion.number;
					todayCache.question = lastquestion.title;
					msg.quesNum = todayCache.quesNum;
					msg.question = todayCache.question;
					console.log(msg, 'request');
					_this.syncLocal();
					if (panel) {
						panel.webview.postMessage(msg);
					}
				}).catch((e: Error) => {
					if (panel) {
						panel.webview.postMessage(msg);
					}
				})
			}
		} else {
			if (panel) {
				panel.webview.postMessage(msg);
			}
		}
		
	}
	dispose() {
		this.statusBar.dispose();
	}
}

export default DailyLine;